### PR TITLE
Python 3 support and get/set item magic methods on DMXConnection.

### PR DESCRIPTION
**Pull request by willstott101**
I've started using `array.array` and `int.to_bytes` to handle integer -> bytes conversions rather than `chr` and `bytes.join`. Since we're now using a byte array for the `dmx_frame` it's impossible to try and send out of range values.

All these changes are backwards compatible.

Thanks for this library!

### CODE REVIEW

**Comment by generalelectrix** [comment]
Thanks for the patch!  Looks good, thanks for tidying up the exception hierarchy.  Would you be willing to add a couple of unit tests of the offline port that exercise the new overflow catching behavior?

**Comment by willstott101** [comment]
To be honest I haven't done a huge amount of automated testing, and it looks as though there are no tests in this repo yet. I'm happy to have a go at adding some in though. Is there any particular library you'd like me to use? https://docs.pytest.org/en/latest/ This looks fairly neat.

**Comment by willstott101** [comment]
Thanks @croadfeldt for noticing that

**Comment by generalelectrix** [comment]
I just updated the package structure to add a test module and a single test of the offline port.  Please merge master into your branch, fix conflicts, then add the test case as a function in test.py

The nose test runner picks up any function whose name starts with `test`.

**Comment by willstott101** [comment]
Well, looks like I completely overlooked the to_bytes thing. Could have sworn I'd tested on python2, but there we go. Another demonstration of the necessity for writing tests.

File: pyenttec/__init__.py [hunk 1]
```diff
@@ -209,10 +213,25 @@ def set_channel(self, chan, val):
             raise DMXAddressError("Channel index {} out of range. "
                                   "Universe size is {}."
                                   .format(chan, len(self.dmx_frame)))
+        except OverflowError:
+            raise DMXOverflowError("Channel value {} out of range. "
+                                  "DMX uses 8bit unsigned values (0-255)."
+                                  .format(chan, len(self.dmx_frame)))
```
Comment (generalelectrix) [inline pyenttec/__init__.py]: Looks like this format string only has one argument but two args have been passed to format, I'm assuming this is a copy/paste error and you meant to include both the channel and the value in the message.

File: pyenttec/__init__.py [hunk 2]
```diff
@@ -209,10 +213,25 @@ def set_channel(self, chan, val):
...
+
+    def __setitem__(self, chan, val):
+        self.set_channel(chan, val)
```
Comment (generalelectrix) [inline pyenttec/__init__.py]: Since in a sense `__setitem__` is the basic implementation here, it probably makes more sense to move the guts of `set_channel` in here and reduce the impl of `set_channel` to
```python
def set_channel(self, chan, val):
    self[chan] = val
```

That avoids the method call overhead for each channel set operation in what is probably your most common case.

File: pyenttec/__init__.py [hunk 3]
```diff
@@ -235,11 +254,20 @@ def close(self):
...
+    pass
+
+class EnttecConfigError(EnttecError):
+    pass
+
+class DMXError(Exception):
     pass
 
-class EnttecConfigError(Exception):
+class DMXAddressError(DMXError):
```
Comment (generalelectrix) [inline pyenttec/__init__.py]: I'm not sure that we need this complex exception hierarchy (my opinions here have shifted since writing this library).  `DMXAddressError` is basically just `IndexError`, `DMXOverflowError` is `ValueError`.  Does your use case actually use these exception types in normal operation?

File: pyenttec/test.py [hunk 4]
```diff
@@ -1,15 +1,48 @@
 """Tests for offline DMX port."""
-from nose.tools import assert_equal, assert_raises
-from . import DMXConnectionOffline, DMXAddressError
+from nose.tools import assert_equal, assert_raises, raises
+from . import DMXConnectionOffline, DMXAddressError, DMXOverflowError
+
+def get_test_port(univ_size=24):
```
Comment (generalelectrix) [inline pyenttec/test.py]: All this function does is shadow the offline port constructor, and the kwarg default isn't used in any of the tests.  Since this is a very simple test fixture, please just call the constructor directly.

File: pyenttec/test.py [hunk 5]
```diff
@@ -1,15 +1,48 @@
...
     assert_equal(1, port.dmx_frame[0])
 
-    assert_raises(DMXAddressError, port.set_channel, univ_size, 0)
\ No newline at end of file
+    assert_raises(DMXAddressError, port.set_channel, univ_size, 0)
+
+def test_port_item_accessors():
+    univ_size = 24
+    port = get_test_port(univ_size)
+    port[univ_size - 1] = 2
```
Comment (generalelectrix) [inline pyenttec/test.py]: Very easy here to roll a test that exercises the API harder; in the same LOC count:

```python
for i in range(univ_size):
    port[i] = i
    assert_equal(i, port.dmx_frame[i])
    assert_equal(i, port[i])
```

File: pyenttec/test.py [hunk 6]
```diff
@@ -1,15 +1,48 @@
...
+    assert_equal(2, port.dmx_frame[univ_size - 1])
+    port[0] = 2
+    assert_equal(2, port.dmx_frame[0])
+
+@raises(DMXAddressError)
+def test_port_setitem_max_address():
```
Comment (generalelectrix) [inline pyenttec/test.py]: Also nicer to test more edge cases when probing for them.  Give this a try:

```python
def test_port_setitem_max_address():
    univ_size = 24
    port = get_test_port(univ_size)
    def assert_fail_set_index(i):
        with assert_raises(DMXAddressError):
            port[i] = 0
    assert_fail_set_index(-1)
    assert_fail_set_index(univ_size)
    assert_fail_set_index(univ_size + 1)
```

File: pyenttec/test.py [hunk 7]
```diff
@@ -1,15 +1,48 @@
...
+    univ_size = 24
+    port = get_test_port(univ_size)
+    port[univ_size] = 2
+
+def test_dmx_value_range():
+    port = get_test_port()
+
+    assert_raises(DMXOverflowError, port.set_channel, 1, -1)
```
Comment (generalelectrix) [inline pyenttec/test.py]: This test makes it clear that the exception `DMXOverflowError` is not named well; this is an underflow!  I still think `ValueError` is fine here.

File: pyenttec/test.py [hunk 8]
```diff
@@ -1,15 +1,48 @@
...
+
+    port.set_channel(0, 0)
```
Comment (generalelectrix) [inline pyenttec/test.py]: I like to roll quick helper functions to keep things slim and more easily refactorable:

```python
def set_and_check(chan, val):
    port.set_channel(chan, val)
    assert_equal(val, port[chan])

set_and_check(0, 0)
set_and_check(0, 128)
set_and_check(0, 255)
```

**Review by generalelectrix — CHANGES_REQUESTED** [review]
Thanks for the update; please feel free to pester me if I've forgotten to check in on this.

File: pyenttec/__init__.py (hunk 1 above)
Comment (willstott101) [reply pyenttec/__init__.py]: well spotted

File: pyenttec/__init__.py (hunk 2 above)
Comment (willstott101) [reply pyenttec/__init__.py]: In-fact we could actually do `__setitem__ = set_channel` or the inverse to avoid the overhead in either case. Didn't do that originally cause I think it's less clear, but if we're gonna worry about reducing method calls that's probably a better way to do it.

File: pyenttec/__init__.py (hunk 3 above)
Comment (willstott101) [reply pyenttec/__init__.py]: I think I agree this is unnecessary. But I'm not sure I agree about `ValueError`, I'd be more inclined to use `OverflowError`. It's what I'm catching to replace with DMXOverflowError anyway.

File: pyenttec/test.py (hunk 5 above)
Comment (willstott101) [reply pyenttec/test.py]: Sure, good point

File: pyenttec/test.py (hunk 7 above)
Comment (willstott101) [reply pyenttec/test.py]: Eh, something can overflow on either side can't it?
```python
>>> array.array('B', (-1,))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: unsigned byte integer is less than minimum
```

File: pyenttec/__init__.py (hunk 2 above)
Comment (willstott101) [reply pyenttec/__init__.py]: I'm of the belief that the more descriptive method name should contain the logic, and that `__setitem__` is more of a shortcut. So will push with the `__setitem__ = set_channel` optimisation for now. Can overrule of course if you disagree

File: pyenttec/test.py (hunk 6 above)
Comment (willstott101) [reply pyenttec/test.py]: I like the closure thing, but -1 is a valid index currently. I'm guessing we want to leave as is?

I can't actually see a reason for someone wanting to do that in the context of DMX... but I feel like it's pretty standard python functionality so it might just be less surprising to leave it as allowed.

I guess we could add a test to formalise it.

Comment (chrismacklin) [reply pyenttec/test.py]: Yep, sorry, this feedback was all pre-coffee this morning and of course -1 is entirely valid here.

**Comment by willstott101** [comment]
Thanks for the review! I think I've addressed everything there. I've definitely got a lot of learning to do with Test styles.

File: pyenttec/__init__.py (hunk 2 above)
Comment (chrismacklin) [reply pyenttec/__init__.py]: `set_channel` is a remnant of the API that I forked originally; I don't think it's necessary at all but I don't want to remove it for backwards-compatibility.  Assigning to the attribute looks good to me.

File: pyenttec/__init__.py [hunk 9]
```diff
@@ -183,36 +189,36 @@ def _update_params(self):
...
                         0]
-        self._packet_start = ''.join(chr(v) for v in packet_start)
+        self._packet_start = array('B', packet_start).tostring()
 
         self._write_settings()
 
     def _write_settings(self):
         """Write the current settings to the port."""
-        self.com.write(self._port_params.to_packet())
+        self.com.write(self._port_params.to_packet() + _PACKET_END)
```
Comment (chrismacklin) [inline pyenttec/__init__.py]: Good catch on the missing _PACKET_END; I think it belongs in the `to_packet` method itself, though.

File: pyenttec/__init__.py (hunk 3 above)
Comment (chrismacklin) [reply pyenttec/__init__.py]: `ValueError` seems more appropriate here; the fact that the mechanism by which this value is out of range is overflow is an implementation detail.  From the docs:

`ValueError`: "Raised when a built-in operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError."

`OverflowError`: "Raised when the result of an arithmetic operation is too large to be represented. This cannot occur for integers (which would rather raise MemoryError than give up). However, for historical reasons, OverflowError is sometimes raised for integers that are outside a required range. Because of the lack of standardization of floating point exception handling in C, most floating point operations are not checked."

I've never encountered `OverflowError` even once in regular Python programming and I'd be surprised if setting a value failed with it.  On the other hand, I'd expect `ValueError` for a value outside of the expected range.

**Comment by generalelectrix** [comment]
Sorry for the comments from multiple accounts, I jumped onto a different computer and forgot to swap accounts.

File: pyenttec/__init__.py (hunk 3 above)
Comment (willstott101) [reply pyenttec/__init__.py]: Point taken, I'll stick with ValueError

File: pyenttec/__init__.py (hunk 9 above)
Comment (willstott101) [reply pyenttec/__init__.py]: Oops! In-fact I don't think it is so great. `to_packet` already has `_END_VAL` added onto it's array. Perhaps I will change it to use `_PACKET_END` in that method to be consistent.

**Comment by willstott101** [comment]
Ahah, same person. No trouble. Looking at that `to_packet` implementation again inspired me to switch to tuples when building packets. I'm trying to learn to use tuples more often in general. It's slightly more consistent now with `array.tostring() + _PACKET_END` everywhere.

**Comment by generalelectrix** [comment]
FYI I've had a busy week and I'm about to leave for the weekend; early next week I'll make sure everything still works with my existing projects and get this merged.  Thanks again for the patch!

**Comment by willstott101** [comment]
No problem! Take your time.

**Comment by willstott101** [comment]
Was trying to better handle hot-plugging the enttec in my use-case. Was convenient to be able to store the dmx_frame in my wrapper and continue writing to it until re-connection

**Comment by generalelectrix** [comment]
Hey, sorry I forgot about this MR.  Just merged it down to master, I'll push a new version to PyPI.  Thanks for the patch!
